api/src/utils/3d-network.js

```diff
@@ -0,0 +1,73 @@
+import createGraph from 'ngraph.graph';
+import createLayout from 'ngraph.forcelayout';
+
+const SCALE = 5;
+const MAX_ITERATIONS = 1000;
+
+const populateWithLayout = ({ nodes, links }) => {
+  const g = createGraph();
+
+  for (let node of nodes) {
+    const { id, ...data } = node;
+    g.addNode(id, data);
+  }
+
+  for (let link of links) {
+    const { s, t } = link;
+    g.addLink(s, t);
+  }
+
+  const startTime = Date.now();
+  const layout = createLayout(g, { dimensions: 3 });
+
+  let iterations = MAX_ITERATIONS;
+  const elementsCount = nodes.length + links.length;
+  if (elementsCount > 50000) {
+    // From benchmarks, it takes about 60s to run 300 iterations
+    // for a network of around 60000 elements (nodes + links) 
+    // so the iterations is capped at 200 for optimal speed
+    iterations = 200;
+  } else if (elementsCount > 12000) {
+    // From benchmarks, it takes about 25s and 800 iterations
+    // to stabilize a network of around 11000 elements (nodes + links) 
+    // so the iterations is capped at 800 for optimal quality
+    iterations = 800;
+  }
+
+  for (let i = 1; i <= iterations; ++i) {
+    const isStable = layout.step();
+
+    if (isStable) {
+      break;
+    }
+  }
+
+  const nodeSet = new Set();
+  const linkSet = new Set();
+
+  g.forEachNode((node) => {
+    const { x, y, z } = layout.getNodePosition(node.id);
+    const pos = [
+      Math.round(x * SCALE),
+      Math.round(y * SCALE),
+      Math.round(z * SCALE),
+    ];
+
+    nodeSet.add({
+      id: node.id,
+      pos,
+      ...node.data,
+    });
+
+    for (let link of node.links) {
+      linkSet.add({ s: link.fromId, t: link.toId });
```

Comment:
I think I noticed duplicates for both nodes and and links out of cypher. After the talk yesterday I have a better idea of why they are there now. Using sets just makes sure there are no duplicates.

Reply:
Would there be a performance penalty if the issue would be fixed at the source (cypher)?

Reply:
Hmm I think I had problems with doing this in cypher but we could probably convert them into sets before using this utility function. I imagine that could also lead to more efficient stabilization. Let me try it. Thanks!

Reply:
So.. a line of interesting observations:
- Actually, when I was looking into this now I see that the nodes and links are already unique at line 7. I think I was using a `Set` before for the graph calculations, links are nested into each relevant node.
- However, it doesn't actually look like the `Set` is working. This is kind of weird because according to the [doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set), it should work just fine for objects. We were actually returning exactly two copies of every link.
- I did realize by now that I could just use the original links, as it is only the nodes that need to be updated anyways. That removes the need for using Sets (not that they are working) and the graph should actually look slightly cleaner without double links everywhere.

Reply:
It's nice that you figured out a simpler way.